sys/drv/plic.c

```diff
@@ -0,0 +1,190 @@
+#define KL_LOG KL_DEV
+#include <sys/bus.h>
+#include <sys/devclass.h>
+#include <sys/errno.h>
+#include <sys/fdt.h>
+#include <sys/interrupt.h>
+#include <sys/klog.h>
+#include <sys/libkern.h>
+
+/* PLIC memory map. */
+#define PLIC_CTXNUM_SV 1
+
+#define PLIC_PRIORITY_BASE 0x000000
+
+#define PLIC_ENABLE_BASE 0x002000
+#define PLIC_ENABLE_STRIDE 0x80
+#define PLIC_ENABLE_BASE_SV                                                    \
+  (PLIC_ENABLE_BASE + PLIC_CTXNUM_SV * PLIC_ENABLE_STRIDE)
+
+#define PLIC_CONTEXT_BASE 0x200000
+#define PLIC_CONTEXT_STRIDE 0x1000
+#define PLIC_CONTEXT_THRESHOLD 0x0
+#define PLIC_CONTEXT_CLAIM 0x4
+#define PLIC_CONTEXT_BASE_SV                                                   \
+  (PLIC_CONTEXT_BASE + PLIC_CTXNUM_SV * PLIC_CONTEXT_STRIDE)
+
+#define PLIC_PRIORITY(n) (PLIC_PRIORITY_BASE + (n) * sizeof(uint32_t))
+#define PLIC_ENABLE_SV(n) (PLIC_ENABLE_BASE_SV + ((n) / 32) * sizeof(uint32_t))
+#define PLIC_THRESHOLD_SV (PLIC_CONTEXT_BASE_SV + PLIC_CONTEXT_THRESHOLD)
+#define PLIC_CLAIM_SV (PLIC_CONTEXT_BASE_SV + PLIC_CONTEXT_CLAIM)
+
+typedef struct plic_state {
+  rman_t rm;                 /* irq resource manager */
+  resource_t *mem;           /* PLIC memory resource */
+  resource_t *irq;           /* PLIC irq resource */
+  intr_event_t **intr_event; /* interrupt events */
+  unsigned ndev;             /* number of sources */
+} plic_state_t;
+
+#define in4(addr) bus_read_4(plic->mem, (addr))
+#define out4(addr, val) bus_write_4(plic->mem, (addr), (val))
+
+static void plic_intr_disable(intr_event_t *ie) {
+  plic_state_t *plic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  uint32_t en = in4(PLIC_ENABLE_SV(irq));
+  en &= ~(1 << (irq % 32));
+  out4(PLIC_ENABLE_SV(irq), en);
+}
+
+static void plic_intr_enable(intr_event_t *ie) {
+  plic_state_t *plic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  uint32_t en = in4(PLIC_ENABLE_SV(irq));
+  en |= 1 << (irq % 32);
+  out4(PLIC_ENABLE_SV(irq), en);
+}
+
+static const char *plic_intr_name(unsigned irq) {
+  char buf[32];
+  snprintf(buf, sizeof(buf), "PLIC source %u", irq);
+  return kstrndup(M_STR, buf, sizeof(buf));
```

Comment:
Is there a chance you will add `kasprintf` function to the kernel?

Reply:
#1251. Done.